### PR TITLE
Add TT21100 touchscreen component

### PR DESCRIPTION
**Pull request by kroimon**
# What does this implement/fix?

Implements a component to interface with the TT21100 touchscreen.

Besides others, this chips is used on the Espressif ESP32-S3-BOX and ESP32-S3-Korvo-2-LCD. Other development efforts regarding this hardware is being coordinated [here](https://github.com/esphome/feature-requests/issues/2239).

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#2916

## Test Environment

- [X] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:
```yaml
i2c:
  scl: GPIO18
  sda: GPIO8

touchscreen:
  - platform: tt21100
    interrupt_pin: GPIO3
    reset_pin: GPIO48

binary_sensor:
  - platform: tt21100
    name: "Home"
    index: 0
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by kroimon** [comment]
For anyone wanting to test this PR on an ESP32-S3-BOX, [here is the configuration YAML](https://gist.github.com/kroimon/f6692879f9c00702990801ae9dfa433b) I used.
Buttons, Display and Touchscreen are working.

**Comment by kroimon** [comment]
I have separated the buttons out into their own component as suggested by @ssieb and added documentation in esphome/esphome-docs#2916.

This is now ready for review.

**Comment by gsgxnet** [comment]
I tried the PR with a config as stated here: 


https://gist.github.com/kroimon/f6692879f9c00702990801ae9dfa433b?permalink_comment_id=4607767#gistcomment-4607767

All okay

**Comment by rpatel3001** [comment]
Could you have the states of the buttons reported at poweron? Red/green/blue stay as unknown until the screen is touched (even not in a button zone). And for some reason "home" doesn't update until it's specifically touched.

**Comment by rpatel3001** [comment]
added initial state in https://github.com/kroimon/esphome/pull/88

File: esphome/components/tt21100/binary_sensor/__init__.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import esphome.codegen as cg
+import esphome.config_validation as cv
+from esphome.components import binary_sensor
+from esphome.const import CONF_INDEX
+
+from .. import tt21100_ns
+from ..touchscreen import TT21100Touchscreen, TT21100ButtonListener
+
+CONF_TT21100_ID = "tt21100_id"
+
```
Comment (jesserockz) [inline esphome/components/tt21100/binary_sensor/__init__.py:10]:
```suggestion
DEPENDENCIES = ["tt21100"]

```

Comment (jesserockz) [reply esphome/components/tt21100/binary_sensor/__init__.py:10]: This gives a nicer validation error

**Review by jesserockz — DISMISSED** [review]

File: esphome/components/tt21100/binary_sensor/__init__.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
+import esphome.codegen as cg
+import esphome.config_validation as cv
+from esphome.components import binary_sensor
+from esphome.const import CONF_INDEX
+
+from .. import tt21100_ns
+from ..touchscreen import TT21100Touchscreen, TT21100ButtonListener
+
+CONF_TT21100_ID = "tt21100_id"
+DEPENDENCIES = ["tt21100"]
```
Comment (jesserockz) [inline esphome/components/tt21100/binary_sensor/__init__.py]: Oops, forgot it was a touchscreen platform...haha
```suggestion
```

**Pull request merged**